pkg/controller/anomalydetector/controller.go

```diff
@@ -0,0 +1,649 @@
+		anomalyDetectorInformer: taDetectorInformer.Informer(),
+		anomalyDetectorLister:   taDetectorInformer.Lister(),
+		anomalyDetectorSynced:   taDetectorInformer.Informer().HasSynced,
+		periodicResyncSet:       make(map[apimachinerytypes.NamespacedName]struct{}),
+	}
+
+	c.anomalyDetectorInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addTADetector,
+			UpdateFunc: c.updateTADetector,
+			DeleteFunc: c.deleteTADetector,
+		},
+		resyncPeriod,
+	)
+
+	return c
+}
+
+func (c *AnomalyDetectorController) addTADetector(obj interface{}) {
+	newTAD, ok := obj.(*crdv1alpha1.ThroughputAnomalyDetector)
+	if !ok {
+		klog.ErrorS(nil, "fail to convert to Throughput Anomaly Detector", "object", obj)
+		return
+	}
+	klog.V(2).InfoS("Processing Throughput Anomaly Detector ADD event", "name", newTAD.Name, "labels", newTAD.Labels)
+	namespacedName := apimachinerytypes.NamespacedName{
+		Namespace: newTAD.Namespace,
+		Name:      newTAD.Name,
+	}
+	c.queue.Add(namespacedName)
+}
+
+func (c *AnomalyDetectorController) updateTADetector(_, new interface{}) {
+	newTAD, ok := new.(*crdv1alpha1.ThroughputAnomalyDetector)
+	if !ok {
+		klog.ErrorS(nil, "fail to convert to Throughput Anomaly Detector", "object", new)
+		return
+	}
+	klog.V(2).InfoS("Processing Throughput Anomaly Detector UPDATE event", "name", newTAD.Name, "labels", newTAD.Labels)
+	namespacedName := apimachinerytypes.NamespacedName{
+		Namespace: newTAD.Namespace,
+		Name:      newTAD.Name,
+	}
+	c.queue.Add(namespacedName)
+}
+
+func (c *AnomalyDetectorController) deleteTADetector(old interface{}) {
+	newTAD, ok := old.(*crdv1alpha1.ThroughputAnomalyDetector)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Error decoding object when deleting Throughput Anomaly Detector", "oldObject", old)
+			return
+		}
+		newTAD, ok = tombstone.Obj.(*crdv1alpha1.ThroughputAnomalyDetector)
+		if !ok {
+			klog.ErrorS(nil, "Error decoding object tombstone when deleting Throughput Anomaly Detector", "tombstone", tombstone.Obj)
+			return
+		}
+	}
+	klog.V(2).InfoS("Processing Throughput Anomaly Detector DELETE event", "name", newTAD.Name, "labels", newTAD.Labels)
+	// remove Throughput Anomaly Detector from periodic synchronization list in case it is deleted before completing
+	c.stopPeriodicSync(apimachinerytypes.NamespacedName{
+		Namespace: newTAD.Namespace,
+		Name:      newTAD.Name,
+	})
+	// Add SparkApplication and Namespace information to deletionQueue for cleanup
+	if newTAD.Status.SparkApplication != "" {
+		namespacedId := NamespacedId{
+			Namespace: newTAD.Namespace,
+			Id:        newTAD.Status.SparkApplication,
+		}
+		c.deletionQueue.Add(namespacedId)
+	}
+}
+
+// Run will create defaultWorkers workers (go routines) which will process the Service events from the
+// workqueue.
+func (c *AnomalyDetectorController) Run(stopCh <-chan struct{}) {
```

Comment:
We may also need a resync logic like we have here #150  for NPR. And maybe reformat or reuse some of the code. But I'm fine to do it in a follow up PR.

Reply:
Sure, I would like to defer it to https://github.com/antrea-io/theia/issues/168